From 9b4e63947666b31d58a839353a5d88de2d9ed0ee Mon Sep 17 00:00:00 2001
From: "Wang, ArvinX" <arvinx.wang@intel.com>
Date: Fri, 26 Jul 2019 15:12:28 +0530
Subject: [PATCH] Fix testScreenBrightness failed in CtsStatsdHostTestCases 
 Module

Case fail due to update a wrong backlight brightness to the Setting database.
Due to the display brightness for the vehicle, need a percent brightness.
but the System screen backlight brightness is a value between 0 and 255, need
to convert the screen backlight brightness to the percent brightness. If the
screen backlight brightness had changed from the vehicle, need to convert the
percent brightness to screen backlight brightness value, So unable to save a
accurate brightness value to the Setting database.

Test:
1. adb shell settings put system screen_brightness 100
2. adb shell settings get system screen_brightness

Change-Id: Ib4f387c123459357220c7bd74a3df7997c6ca231
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-73451
Signed-off-by: Wang, ArvinX <arvinx.wang@intel.com>
---
 .../car/systeminterface/DisplayInterface.java | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/service/src/com/android/car/systeminterface/DisplayInterface.java b/service/src/com/android/car/systeminterface/DisplayInterface.java
index 8dc7bd40..13f0cb0b 100644
--- a/service/src/com/android/car/systeminterface/DisplayInterface.java
+++ b/service/src/com/android/car/systeminterface/DisplayInterface.java
@@ -125,8 +125,28 @@ public interface DisplayInterface {
             return disp.getState() == Display.STATE_ON;
         }
 
+        private static int brightnessToPercent(int brightness, int minimumBacklight, int maximumBacklight) {
+            int range = maximumBacklight - minimumBacklight;
+            // Convert brightness from minimumBacklight-maximumBacklight to 0-100%
+            brightness -= minimumBacklight;
+            brightness *= 100;
+            brightness += (range + 1) / 2;
+            return brightness /= range;
+        }
+
         @Override
         public void setDisplayBrightness(int percentBright) {
+	    try {
+                int old_brightness = System.getInt(mContentResolver, System.SCREEN_BRIGHTNESS);
+                // Convert brightness from minimumBacklight-maximumBacklight to 0-100%
+                old_brightness = brightnessToPercent(old_brightness, mMinimumBacklight, mMaximumBacklight);
+                // If the new brightness is same with old_brightness, ignore this change.
+                if (old_brightness == percentBright) {
+                    return;
+                }
+            } catch (SettingNotFoundException e) {
+                Log.e(CarLog.TAG_POWER, "Could not get SCREEN_BRIGHTNESS:  " + e);
+            }
             int gamma = (percentBright * GAMMA_SPACE_MAX + 50) / 100;
             int linear = convertGammaToLinear(gamma, mMinimumBacklight, mMaximumBacklight);
             System.putInt(mContentResolver, System.SCREEN_BRIGHTNESS, linear);
-- 
2.22.0

